File: apps/windows/TeslaSync.App/shared-surfaces/AIThinkingIndicator.Model.cs
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata for the AIThinkingIndicator surface — the native analogue of the module-level
/// constants and default <c>t()</c> calls in <c>web/src/components/ai/AIThinkingIndicator.tsx</c>. The web
/// component is a pure presentational "streaming-but-empty" indicator (the state shown while the SSE
/// connection is open and the first <c>delta.text</c> frame has not arrived), so this carries the diagnostics
/// slug, the <c>data-testid</c> automation id, the two i18n keys the source references (the default
/// <c>helix.thinking</c> label and the <c>ai.common.thinking</c> override example from the JSDoc), and the
/// structural constants the view composes from (three prose skeleton lines whose decreasing widths mimic text,
/// and three trailing "thinking" dots).
/// </summary>
public static class AIThinkingIndicatorRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "AIThinkingIndicator";

    /// <summary>
    /// The root automation id — the native analogue of the web <c>data-testid="ai-thinking-indicator"</c>.
    /// </summary>
    public const string RootAutomationId = "ai-thinking-indicator";

    /// <summary>
    /// i18n key for the default leading label (web <c>t('helix.thinking', 'Helix is thinking')</c>). The
    /// resource catalog is <c>translation</c>-namespaced, so the web key <c>helix.thinking</c> resolves under
    /// <see cref="HelixThinkingKey"/> here; absent a catalog entry the i18n facade returns
    /// <see cref="HelixThinkingFallback"/> verbatim, exactly as the peer AI surfaces do.
    /// </summary>
    public const string HelixThinkingKey = "translation.helix.thinking";

    /// <summary>English fallback for <see cref="HelixThinkingKey"/> (web second arg, verbatim — no ellipsis).</summary>
    public const string HelixThinkingFallback = "Helix is thinking";

    /// <summary>
    /// i18n key for the generic AI thinking label referenced by the web source's JSDoc as the override a
    /// summary-style surface would pass (web <c>t('ai.common.thinking', 'AI is thinking')</c>). Exposed so a
    /// host can request the generic verb instead of the Helix-branded default.
    /// </summary>
    public const string AiCommonThinkingKey = "translation.ai.common.thinking";

    /// <summary>English fallback for <see cref="AiCommonThinkingKey"/> (web second arg, verbatim).</summary>
    public const string AiCommonThinkingFallback = "AI is thinking";

    /// <summary>
    /// The ARIA role the web container declares (<c>role="status"</c>) — surfaced so the view and tests agree on
    /// the live-region contract the screen reader observes.
    /// </summary>
    public const string Role = "status";

    /// <summary>The ARIA live urgency the web container declares (<c>aria-live="polite"</c>).</summary>
    public const string LiveSetting = "polite";

    /// <summary>The number of trailing "thinking" dots the web indicator renders.</summary>
    public const int DotCount = 3;

    /// <summary>
    /// The decreasing widths of the three prose skeleton lines, as (numerator, denominator) fractions of the
    /// container width — the native analogue of the web <c>w-full</c>, <c>w-11/12</c> and <c>w-9/12</c> classes.
    /// Kept as exact fractions so the view can lay them out proportionally (responsive) rather than at a fixed
    /// pixel width.
    /// </summary>
    public static IReadOnlyList<(int Numerator, int Denominator)> SkeletonLineFractions { get; } =
        new (int, int)[] { (1, 1), (11, 12), (9, 12) };
}

/// <summary>
/// Pure projection of the indicator's render inputs — the native port of the web component body's two
/// decisions (web/src/components/ai/AIThinkingIndicator.tsx): the resolved leading label
/// (<c>label ?? t('helix.thinking', 'Helix is thinking')</c>) and whether motion runs (the web
/// <c>motion-safe:</c> variant, which drops the bouncing dots and line shimmer under
/// <c>prefers-reduced-motion</c>). Kept static and side-effect-free so the adapter is unit-testable without a
/// view-model or a UI thread; the <see cref="AIThinkingIndicatorViewModel"/> and the WinUI view both render
/// from it.
/// </summary>
public readonly record struct AIThinkingProjection
{
    private AIThinkingProjection(string label, bool animate)
    {
        Label = label;
        Animate = animate;
        Role = AIThinkingIndicatorRegistration.Role;
        LiveSetting = AIThinkingIndicatorRegistration.LiveSetting;
    }

    /// <summary>
    /// The resolved leading label — the screen-reader-visible status text. This is also the surface's
    /// accessible name, mirroring the web <c>role="status"</c> element whose only non-hidden content is the
    /// label (the HelixMark, dots and skeleton lines are <c>aria-hidden</c>).
    /// </summary>
    public string Label { get; }

    /// <summary>
    /// Whether the indicator animates (bouncing dots + shimmering lines). False under reduced motion, where the
    /// static skeleton is still shown — the web <c>motion-safe:</c> behaviour. Equivalent to
    /// <c>MotionDuration.ShouldAnimate(reduceMotion)</c>.
    /// </summary>
    public bool Animate { get; }

    /// <summary>The ARIA role the surface exposes (always <see cref="AIThinkingIndicatorRegistration.Role"/>).</summary>
    public string Role { get; }

    /// <summary>The ARIA live urgency the surface exposes (always <see cref="AIThinkingIndicatorRegistration.LiveSetting"/>).</summary>
    public string LiveSetting { get; }

    /// <summary>
    /// Project the render inputs. When <paramref name="customLabel"/> is null/blank the default label is
    /// resolved through the i18n facade (web <c>t('helix.thinking', 'Helix is thinking')</c>); otherwise the
    /// caller's already-translated override is used verbatim (web <c>label ?? ...</c>). <paramref name="reduceMotion"/>
    /// flips <see cref="Animate"/> off (web <c>prefers-reduced-motion</c>).
    /// </summary>
    /// <param name="localizer">The i18n facade the default label resolves through.</param>
    /// <param name="customLabel">An optional already-translated override label.</param>
    /// <param name="reduceMotion">Whether the OS reduce-motion preference is set.</param>
    public static AIThinkingProjection Project(ILocalizer localizer, string? customLabel, bool reduceMotion)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        var label = string.IsNullOrWhiteSpace(customLabel)
            ? localizer.GetString(
                AIThinkingIndicatorRegistration.HelixThinkingKey,
                AIThinkingIndicatorRegistration.HelixThinkingFallback)
            : customLabel;

        return new AIThinkingProjection(label, animate: !reduceMotion);
    }
}

/// <summary>
/// PII-safe diagnostics for the AIThinkingIndicator surface (P1/S11 diagnostics contract). The indicator
/// carries no user content (only a static "thinking" label), so the collector records the operational
/// <c>view.opened</c> event with the surface slug and nothing else. Thread-safe.
/// </summary>
public sealed class AIThinkingIndicatorDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public AIThinkingIndicatorDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=AIThinkingIndicator</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={AIThinkingIndicatorRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/AIThinkingIndicator/AIThinkingIndicatorModel.kt
// Pure, framework-free model + render projection + diagnostics for the AIThinkingIndicator shared surface —
// the native analogue of everything the web component derives (web/src/components/ai/AIThinkingIndicator.tsx).
// No Compose, no Android framework, no HTTP: every declaration here is exercised off-device in the
// :android:testReleaseUnitTest gate, keeping the composable a thin render layer.
//
// The web source is a STATIC presentational surface — the streaming-but-empty pending state shown while an AI
// surface holds an open SSE connection and waits for the first `delta.text` frame (it is what `AiOutputPanel`
// renders as its `pendingChild`). It draws an animated "Helix is thinking" label (a HelixMark glyph + bouncing
// dots) over three shimmering skeleton lines of decreasing width. Its only inputs are:
//   • `label` — an optional caller override (web `label ?? t('helix.thinking', 'Helix is thinking')`).
//   • `prefers-reduced-motion` — the web `motion-safe:` variant gates the bounce + shimmer; under reduced motion
//     the dots stop bouncing and the lines drop the shimmer, while the static skeleton stays visible.
// There is no data fetch and no second exported "Dots" form beyond the compact in-button variant.
//
// Parity-with-honesty (Honesty Covenant #9 — documented, not silent): the generic
// loading / empty / error / stale / offline data-states do NOT apply to this surface. AIThinkingIndicator IS
// itself the "loading" state of a host AI surface — it has no feed of its own to be empty, stale, offline, or to
// fail, so modelling those would fabricate behaviour the web spec does not have (the same rationale the accepted
// AIChatbotIndicator / VisuallyHidden ports document). The surface's real, reproduced states are the animated
// indicator (full motion) and its reduced-motion variant (static skeleton, no bounce/shimmer), plus the caller's
// label override and the compact dots form — all projected purely here and unit-tested off-device.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/AIThinkingIndicator — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling AIChatbotIndicator / VisuallyHidden surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.aithinkingindicator

import io.teslasync.shared.core.diagnostics.Logger

/**
 * Diagnostics surface slug emitted with the one-shot `view.opened` event (P1/S11). It is the surface slug the
 * prompt mandates (`AIThinkingIndicator`) and carries no VIN, vehicle id, label text, or any model output, so a
 * diagnostics line can never leak the operator's fleet state or what the AI was asked.
 */
const val AI_THINKING_INDICATOR_SLUG: String = "AIThinkingIndicator"

/** The stable, dot-namespaced diagnostics event emitted once when the surface first composes (P1/S11). */
const val EVENT_VIEW_OPENED: String = "view.opened"

/** The structured-field key carrying the surface slug on the `view.opened` diagnostic. */
const val FIELD_SURFACE: String = "surface"

/**
 * The web catalog key whose value is the canonical "Helix is thinking…" string (web `chatbot.thinking`, present
 * in the P1/S10 catalog as `R.string.translation_chatbot_thinking`). The web source's default label
 * `t('helix.thinking', 'Helix is thinking')` resolves to this exact string at runtime — `helix.thinking` is an
 * i18next key with no catalog entry, so i18next returns its fallback, which the catalog already canonicalises
 * under `chatbot.thinking`. The composable binds the catalog resource as the default; this constant records the
 * provenance so the mapping is auditable (Honesty Covenant #9).
 */
const val DEFAULT_LABEL_CATALOG_KEY: String = "chatbot.thinking"

/** Web `w-full` — the first skeleton line fills the full width. */
const val SKELETON_FRACTION_FULL: Float = 1f

/** Web `w-11/12` — the second skeleton line. */
const val SKELETON_FRACTION_ELEVEN_TWELFTHS: Float = 11f / 12f

/** Web `w-9/12` — the third (shortest) skeleton line, mimicking the ragged end of a paragraph. */
const val SKELETON_FRACTION_THREE_QUARTERS: Float = 9f / 12f

/** Web per-line shimmer offset (`[animation-delay:0.3s]` / `0.6s`) — each line trails the previous by this much. */
const val SKELETON_LINE_STAGGER_MS: Int = 300

/**
 * Web per-dot bounce offset. The web dots use negative delays (`-0.3s` / `-0.15s` / `0`) so they start mid-cycle
 * and ripple left-to-right; the native renderer reproduces that ripple by fast-forwarding each dot by this step.
 */
const val THINKING_DOT_STAGGER_MS: Int = 150

/** The web indicator draws exactly three skeleton lines and three dots. */
const val THINKING_DOT_COUNT: Int = 3

/**
 * The immutable surface state the [AIThinkingIndicatorViewModel] exposes. It carries only the resolved
 * reduced-motion preference (web `prefers-reduced-motion`, read through the `motion-safe:` variant); the label is
 * a pure render parameter threaded straight through (web's `label` prop), so it is not part of the bound state.
 *
 * @property reducedMotion whether the platform requests reduced motion. `true` stops the dot bounce and the line
 *   shimmer while keeping the static skeleton visible (web `motion-safe:` → animation suppressed). Defaults to
 *   `false` (full motion) so the surface animates until the preference resolves.
 */
data class ThinkingIndicatorState(
    val reducedMotion: Boolean = false,
)

/**
 * One shimmering skeleton line: its [widthFraction] of the parent width (web `w-full` / `w-11/12` / `w-9/12`) and
 * the [animationDelayMs] by which its shimmer trails the line above (web `[animation-delay:…]`). The delay is
 * inert under reduced motion (the line renders as a static bar).
 */
data class SkeletonLineSpec(
    val widthFraction: Float,
    val animationDelayMs: Int,
)

/**
 * One bouncing dot: the [animationDelayMs] (a fast-forward offset) that desynchronises it from its neighbours so
 * the three dots ripple (web's negative `[animation-delay]`). Inert under reduced motion (the dot renders static).
 */
data class ThinkingDotSpec(
    val animationDelayMs: Int,
)

/**
 * The render-ready projection of the full indicator: the resolved [label], whether the dots/lines should animate
 * ([animated] — the inverse of reduced motion), and the fixed [lines] + [dots] geometry the web encodes. A closed
 * value the stateless composable consumes directly, so the render path branches on nothing else and every field
 * is unit-tested off-device.
 */
data class ThinkingIndicatorProjection(
    val label: String,
    val animated: Boolean,
    val lines: List<SkeletonLineSpec>,
    val dots: List<ThinkingDotSpec>,
)

/**
 * The render-ready projection of the compact [AIThinkingDots] variant — whether the dots bounce ([animated]) and
 * the fixed [dots] geometry. The compact form carries no skeleton lines and no HelixMark; it is the in-button
 * streaming label the web exports alongside the full indicator.
 */
data class ThinkingDotsProjection(
    val animated: Boolean,
    val dots: List<ThinkingDotSpec>,
)

/**
 * Resolves the label the indicator shows — the pure port of web `label ?? t('helix.thinking', …)`. A non-null
 * [override] (an already-localized, caller-supplied string such as "Helix is summarising") wins; otherwise the
 * catalog [default] ("Helix is thinking…") is used. Nullish semantics match the web `??`: an explicitly empty
 * override is the caller's choice and is passed through unchanged.
 */
fun resolveThinkingLabel(
    override: String?,
    default: String,
): String = override ?: default

/**
 * The three skeleton lines, top-to-bottom, with their decreasing widths and trailing shimmer offsets — the pure
 * port of the web `flex flex-col gap-2` block. Stable order, so the renderer and the tests agree on every line.
 */
fun thinkingSkeletonLines(): List<SkeletonLineSpec> =
    listOf(
        SkeletonLineSpec(SKELETON_FRACTION_FULL, animationDelayMs = 0),
        SkeletonLineSpec(SKELETON_FRACTION_ELEVEN_TWELFTHS, animationDelayMs = SKELETON_LINE_STAGGER_MS),
        SkeletonLineSpec(SKELETON_FRACTION_THREE_QUARTERS, animationDelayMs = 2 * SKELETON_LINE_STAGGER_MS),
    )

/**
 * The three bouncing dots, left-to-right, each fast-forwarded one [THINKING_DOT_STAGGER_MS] step past the last so
 * they ripple — the pure port of the web dot row's staggered negative delays. Shared by the full indicator and
 * the compact [AIThinkingDots] form.
 */
fun thinkingDots(): List<ThinkingDotSpec> =
    (0 until THINKING_DOT_COUNT).map { index -> ThinkingDotSpec(animationDelayMs = index * THINKING_DOT_STAGGER_MS) }

/**
 * Projects the full indicator for [state] and the caller's [labelOverride] against the catalog [defaultLabel].
 * Pure (no Compose / clock): the label is resolved, motion is enabled only when the platform does not request
 * reduced motion, and the fixed line + dot geometry is attached. The stateless composable renders this verbatim.
 */
fun projectThinkingIndicator(
    state: ThinkingIndicatorState,
    labelOverride: String?,
    defaultLabel: String,
): ThinkingIndicatorProjection =
    ThinkingIndicatorProjection(
        label = resolveThinkingLabel(labelOverride, defaultLabel),
        animated = !state.reducedMotion,
        lines = thinkingSkeletonLines(),
        dots = thinkingDots(),
    )

/**
 * Projects the compact [AIThinkingDots] variant: the dots bounce only when [reducedMotion] is `false`. Pure, so
 * the in-button form's motion gating is unit-tested without a Compose host.
 */
fun projectThinkingDots(reducedMotion: Boolean): ThinkingDotsProjection =
    ThinkingDotsProjection(animated = !reducedMotion, dots = thinkingDots())

/**
 * Emits the one PII-safe `view.opened` diagnostic carrying only the surface [AI_THINKING_INDICATOR_SLUG] (P1/S11)
 * — never the label text, a vehicle id, or any model output, so a diagnostics line can never leak fleet state or
 * the prompt. Kept free of Compose so it is unit-tested with a recording [Logger]; the ViewModel calls it once
 * per surface open.
 */
fun recordThinkingIndicatorOpened(logger: Logger) {
    logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to AI_THINKING_INDICATOR_SLUG))
}
